CRM/Core/Payment.php

```diff
@@ -134,48 +135,149 @@ abstract class CRM_Core_Payment {
   protected $billingProfile;
 
   /**
-   * Payment instrument ID.
+   * Is this a back office transaction.
+   *
+   * @var bool
+   */
+  protected $backOffice = FALSE;
+
+
+  /**
+   * This is only needed during the transitional phase. In future you should
+   * pass your own PropertyBag into the method you're calling.
+   *
+   * New code should NOT use $this->propertyBag.
    *
-   * This is normally retrieved from the payment_processor table.
+   * This replaces inputParams (see __set and __get)
    *
-   * @var int
+   * @var Civi\Payment\PropertyBag
    */
-  protected $paymentInstrumentID;
+  protected $propertyBag;
 
   /**
-   * Is this a back office transaction.
+   * Magic, handle the response now.
+   */
+  public function __isset($prop) {
+    if ($prop === 'inputParams') {
+      // We always have one of these.
+      return TRUE;
+    }
+    elseif ($prop === 'amount') {
+      return $this->getInternalPropertyBag()->has('amount');
+    }
+    elseif ($prop === 'currency') {
+      return $this->getInternalPropertyBag()->has('currency');
+    }
+  }
+
+  /**
+   * Magic getter to handle legacy properties
    *
-   * @var bool
+   * @param mixed $prop
    */
-  protected $backOffice = FALSE;
+  public function __get($prop) {
+    $propertyBag = $this->getInternalPropertyBag();
+
+    if ($prop === 'inputParams') {
+      return $this->getInternalPropertyBag();
+    }
+    elseif ($prop === 'amount') {
+      // Note that this class never had amount defined as a property, but it was referenced.
+      // To mimic old behaviour we return NULL if the
+      // amount is not set, rather than throwing an exception.
+      if ($propertyBag->has('amount')) {
```

Comment:
the old function retrieved amount from params looking up a couple of keys - what will happen now? ie transitionally if we can't count on it being set & it's not accepting 'params' now

Reply:
sounds good. Ideally commit 1 is add the settings bag in a nice form & the minimum amount of code to support it (which might actually be none) & then we can tweak PaymentProcessor.pay to pass in a settings bag rather than an array & a processor in core to use it a bit